M2/Macaulay2/packages/Complexes/CartanEilenberg.m2

```diff
@@ -0,0 +1,420 @@
+    map(B, Bd, H)
+    )
+
+dthVerticalSubbicomplex = method();
+dthVerticalSubbicomplex(ZZ, Bicomplex) := Bicomplex => (d,B) -> (
+    H = hashTable for i in keys B.dd.map list i => (
+	if i_1 > d and i_2==2 then continue
+	else if i_1 >= d+1 and i_2==1 then continue
+	else if i_1==d+1 and i_2==1 then continue
+	else B.dd_i
+	);
+    bicomplex H
+    )
+
+dthVerticalInclusion = method();
+dthVerticalInclusion(ZZ, Bicomplex) := Bicomplex => (d,B) -> (
+    Bd := dthVerticalSubbicomplex(d, B);
+    H := hashTable for i in keys Bd.module list i => id_(B_i);
+    map(B, Bd, H)
+    )
+
+--new type: FilteredBicomplex. The intent is that we can totalize these objects to
+--obtain filtered complexes, then use the SpectralSequences package to construct induced spectral sequences
+FilteredBicomplex = new Type of MutableHashTable
+FilteredBicomplex.synonym = "filtered bicomplex"
+
+--this type is very barebones: the input is just a list of BicomplexMaps
+filteredBicomplex = method()
+filteredBicomplex(List) := FilteredBicomplex => L -> (
+  maps := L;
+  moduleList := for i from 0 to #L-1 list i =>  source L_i;
+  --Z := image map(C, C,0); -- make zero subcomplex as a subcomplex of ambient complex
+  maps' := hashTable for i from 0 to #maps - 1 list i => maps_i;
+  C := new FilteredBicomplex from {
+	   symbol ring => ring source maps#0,
+	-- TODO: rename module to category agnostic term
+           symbol bicomplexes => new HashTable from moduleList,
+	   symbol inclusions => maps',
+           symbol concentration => (0, #maps),
+           symbol cache => new CacheTable
+           };
+   C
+   )
+
+--given a morphism of complexes, this outputs the induced inclusion
+--of the image into the target
+imageInclusion = method()
+imageInclusion(ComplexMap) := ComplexMap => phi -> (
+    D := image phi;
+    canonicalMap(target D.cache.image, D)
+    )
+
+--given two complexes C and D, outputs the induced tensor bicomplex.
+--the totalization of this complex is isomorphic to the tensor product
+tensorBicomplex = method()
+tensorBicomplex(Complex, Complex) := Bicomplex => (C,D) -> (
+    (loC, hiC) := concentration C;
+    (loD, hiD) := concentration D;
+    L1 := new MutableList from {};
+    for i from loC+1 to hiC do (
+	for j from loD to hiD do (
+	    L1#(#L1) = (i,j,1) => C.dd_i**D_j;
+	    );
+	);
+    L2 := new MutableList from {};
+    for i from loC to hiC do (
+	for j from loD+1 to hiD do (
+	    L2#(#L2) = (i,j,2) => C_i**D.dd_j;
+	    );
+	);
+    H := hashTable ((toList L1)|(toList L2));
+    bicomplex H
+    )
+
+
+
+--constructs resolution of i-cycles from resolutions of homology and boundaries
+horseshoeHomologyRes = method()
+horseshoeHomologyRes(Complex, ZZ) := Complex => (C,i) -> (
+    iota = inducedMap(ker C.dd_i, image C.dd_(i+1));
+    p = inducedMap(coker iota, target iota);
+    horseshoeResolution complex {p, iota}
+    )
+
+--not yet written
+-*
+doubleHorseshoe = method()
```

Comment:
I believe that the ``SpectralSequences`` package has enjoy a major revision which I suspect is incompatible with this code.

Reply:
Oh I didn't realize this file (which isn't actually loaded) was put in Complexes. I will exclude the commit that adds this for now.